consensus/hotstuff/committees/cluster_committee.go

```diff
@@ -0,0 +1,115 @@
+package committees
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/consensus/hotstuff/committees/leader"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+// Cluster represents the committee for a cluster of collection nodes. Cluster
+// committees are epoch-scoped.
+//
+// Clusters build blocks on a cluster chain but must obtain identity table
+// information from the main chain. Thus, block ID parameters in this Committee
+// implementation reference blocks on the cluster chain, which in turn reference
+// blocks on the main chain - this implementation manages that translation.
+type Cluster struct {
+	state    protocol.ReadOnlyState
+	payloads storage.ClusterPayloads
+	me       flow.Identifier
+	// pre-computed leader selection for the full lifecycle of the cluster
+	selection *leader.LeaderSelection
+	// a filter that returns all members of the cluster committee allowed to vote
+	clusterMemberFilter flow.IdentityFilter
+	// initial set of cluster members, WITHOUT updated weight
+	initialClusterMembers flow.IdentityList
+}
+
+func NewClusterCommittee(
+	state protocol.ReadOnlyState,
+	payloads storage.ClusterPayloads,
+	cluster protocol.Cluster,
+	epoch protocol.Epoch,
+	me flow.Identifier,
+) (*Cluster, error) {
+
+	selection, err := leader.SelectionForCluster(cluster, epoch)
+	if err != nil {
+		return nil, fmt.Errorf("could not compute leader selection for cluster: %w", err)
+	}
+
+	com := &Cluster{
+		state:                 state,
+		payloads:              payloads,
+		me:                    me,
+		selection:             selection,
+		clusterMemberFilter:   cluster.Members().Selector(),
+		initialClusterMembers: cluster.Members(),
+	}
+	return com, nil
+}
+
+func (c *Cluster) Identities(blockID flow.Identifier, selector flow.IdentityFilter) (flow.IdentityList, error) {
+
+	// first retrieve the cluster block payload
+	payload, err := c.payloads.ByBlockID(blockID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get cluster payload: %w", err)
+	}
+
+	// if a reference block is specified, use that
+	if payload.ReferenceBlockID != flow.ZeroID {
+		identities, err := c.state.AtBlockID(payload.ReferenceBlockID).Identities(filter.And(
+			selector,
+			c.clusterMemberFilter,
+		))
+		return identities, err
+	}
+
+	// otherwise, this is a root block, in which case we use the initial cluster members
```

Comment:
is it possible a block has `ReferenceBlock == ZeroID`, but isn't a root block?

Reply:
It's considered invalid for a non-root block to have a zero reference block. See https://github.com/onflow/flow-go/pull/157/files#diff-22b3183ccf4fc336356b12cb382421afa1032326808287eed2813eabbac8aba1R110

Reply:
if there is such implication, we could do `isRootBlock := payload.ReferenceBlock == ZeroID` to make it explicit, and then `if isRootBlock ...`

Reply:
👍 Reads clearer that way. Updated in https://github.com/onflow/flow-go/pull/157/commits/51400a17b192b257d68953d7cf73c77733b7bba4